demos/dev_mode_key_provisioning/src/aws_dev_mode_key_provisioning.c

```diff
@@ -43,88 +44,503 @@
+                                                  sizeof( xPublicKeyTemplate ) / sizeof( CK_ATTRIBUTE ),
+                                                  pxPublicKeyHandle );
+    }
+    else if( xPublicKeyType == CKK_EC )
+    {
+        CK_BYTE xEcParams[] = pkcs11DER_ENCODED_OID_P256;
+        size_t xLength;
+        CK_BYTE xEcPoint[ 256 ] = { 0 };
+
+        mbedtls_ecdsa_context * pxEcdsaContext;
+        pxEcdsaContext = ( mbedtls_ecdsa_context * ) xMbedPkContext.pk_ctx;
+
+        /* DER encoded EC point. Leave 2 bytes for the tag and length. */
+        lMbedResult = mbedtls_ecp_point_write_binary( &pxEcdsaContext->grp, &pxEcdsaContext->Q,
+                                                      MBEDTLS_ECP_PF_UNCOMPRESSED, &xLength,
+                                                      xEcPoint + 2, sizeof( xEcPoint ) - 2 );
+        xEcPoint[ 0 ] = 0x04; /* Octet string. */
+        xEcPoint[ 1 ] = ( CK_BYTE ) xLength;
+
+        CK_ATTRIBUTE xPublicKeyTemplate[] =
+        {
+            { CKA_CLASS,     &xClass,           sizeof( xClass )                             },
+            { CKA_KEY_TYPE,  &xPublicKeyType,   sizeof( xPublicKeyType )                     },
+            { CKA_TOKEN,     &xTrue,            sizeof( xTrue )                              },
+            { CKA_VERIFY,    &xTrue,            sizeof( xTrue )                              },
+            { CKA_EC_PARAMS, xEcParams,         sizeof( xEcParams )                          },
+            { CKA_EC_POINT,  xEcPoint,          xLength + 2                                  },
+            { CKA_LABEL,     pucPublicKeyLabel, strlen( ( const char * ) pucPublicKeyLabel ) }
+        };
+
+        xResult = pxFunctionList->C_CreateObject( xSession,
+                                                  ( CK_ATTRIBUTE_PTR ) xPublicKeyTemplate,
+                                                  sizeof( xPublicKeyTemplate ) / sizeof( CK_ATTRIBUTE ),
+                                                  pxPublicKeyHandle );
+    }
+    else
+    {
+        xResult = CKR_ATTRIBUTE_VALUE_INVALID;
+        configPRINTF( ( "Invalid key type. Supported options are CKK_RSA and CKK_EC" ) );
+    }
+
+    mbedtls_pk_free( &xMbedPkContext );
+
+    return xResult;
+}
+
+
+/*-----------------------------------------------------------*/
+
+CK_RV xProvisionGenerateKeyPairRSA( CK_SESSION_HANDLE xSession,
+                                    uint8_t * pucPrivateKeyLabel,
+                                    uint8_t * pucPublicKeyLabel,
+                                    CK_OBJECT_HANDLE_PTR pxPrivateKeyHandle,
+                                    CK_OBJECT_HANDLE_PTR pxPublicKeyHandle )
+{
+    CK_RV xResult;
+    CK_MECHANISM xMechanism =
+    {
+        CKM_RSA_PKCS_KEY_PAIR_GEN, NULL_PTR, 0
+    };
+    CK_FUNCTION_LIST_PTR pxFunctionList;
+    CK_ULONG xModulusBits = 2048;
+    CK_BYTE xPublicExponent[] = { 0x01, 0x00, 0x01 };
+    CK_BYTE xSubject[] = { 0x01, 0x02 };
```

Comment:
For the sake of those who come after, could we use some named constants or macros?  I think it's better to avoid "magic" numbers.